include/atomic.h

```diff
@@ -15,8 +15,21 @@
 extern "C" {
 #endif
 
+#if defined(CONFIG_ARC_EX_ATOMIC)
+typedef int atomic_val_t;
+
+struct arc_atomic_t {
+	atomic_val_t val;
+	int ex_lock;
+};
```

Comment:
Not sure how I feel about this, as it's an API change.  There can be code floating around that is (correctly, even) using atomic_val_t values as regular integer rvalues and this would break it.  And I definitely don't like that it's a struct only on some architectures.  If this is the way it has to be done, let's change it to a struct everywhere, maybe with an "arch_part" bit that you can optionally define to contain your extra stuff.

Also: that's a pretty significant memory overhead to have a separate lock for every atomic value.  Can't you share a single lock (or rather a single lock per userspace address space, I guess)?

Reply:
I agree, if we hope it a struct, it should let the arch decides. But it means a lot of changes.  This PR just consider ARC's case.